File: external/boost_1.55.0/libs/python/test/extract.cpp
// Copyright David Abrahams 2002.
// Distributed under the Boost Software License, Version 1.0. (See
// accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)

#include <boost/python/extract.hpp>
#include <boost/python/list.hpp>
#include <boost/python/module.hpp>
#include <boost/python/def.hpp>
#include <boost/python/class.hpp>
#include <boost/python/reference_existing_object.hpp>
#include <boost/python/return_value_policy.hpp>
#include <boost/python/implicit.hpp>
#include <string>
#include <boost/lexical_cast.hpp>
#define BOOST_ENABLE_ASSERT_HANDLER
#include <boost/assert.hpp>
#include "test_class.hpp"

using namespace boost::python;

typedef test_class<> X;

bool extract_bool(object x) { return extract<bool>(x); }

boost::python::list extract_list(object x)
{
    extract<list> get_list((x));

    // Make sure we always have the right idea about whether it's a list
    bool is_list_1 = get_list.check();
    bool is_list_2 = PyObject_IsInstance(x.ptr(), (PyObject*)&PyList_Type);
    if (is_list_1 != is_list_2) {
        throw std::runtime_error("is_list_1 == is_list_2 failure.");
    }
    return get_list();
}

char const* extract_cstring(object x)
{
    return extract<char const*>(x);
}

std::string extract_string(object x)
{
    std::string s = extract<std::string>(x);
    return s;
}

std::string const& extract_string_cref(object x)
{
#if defined(BOOST_MSVC) && BOOST_MSVC <= 1300
# pragma warning(push)
# pragma warning(disable:4172) // msvc lies about returning a reference to temporary
#elif defined(_MSC_VER) && defined(__ICL) && __ICL <= 900
# pragma warning(push)
# pragma warning(disable:473) // intel/win32 does too
#endif 
    
    return extract<std::string const&>(x);
    
#if defined(BOOST_MSVC) && BOOST_MSVC <= 1300 || defined(_MSC_VER) && defined(__ICL) && __ICL <= 800
# pragma warning(pop)
#endif 
}

X extract_X(object x)
{
    return extract<X>(x);
}

X* extract_X_ptr(object x) { return extract<X*>(x); }

X& extract_X_ref(object x)
{
    extract<X&> get_x(x);
    return get_x;
}

int double_X(object n)
{
    extract<X> x(n);
    return x().value() + x().value();
}

bool check_bool(object x) { return extract<bool>(x).check(); }
bool check_list(object x) { return extract<list>(x).check(); }
bool check_cstring(object x) { return extract<char const*>(x).check(); }
bool check_string(object x) { return extract<std::string>(x).check(); }
bool check_string_cref(object x) { return extract<std::string const&>(x).check(); }
bool check_X(object x) { return extract<X>(x).check(); }
bool check_X_ptr(object x) { return extract<X*>(x).check(); }
bool check_X_ref(object x) { return extract<X&>(x).check(); }

std::string x_rep(X const& x)
{
    return "X("  + boost::lexical_cast<std::string>(x.value()) + ")";
}

BOOST_PYTHON_MODULE(extract_ext)
{
    implicitly_convertible<int, X>();

    def("extract_bool", extract_bool);
    def("extract_list", extract_list);
    def("extract_cstring", extract_cstring);
    def("extract_string", extract_string);
    def("extract_string_cref", extract_string_cref, return_value_policy<reference_existing_object>());
    def("extract_X", extract_X);
    def("extract_X_ptr", extract_X_ptr, return_value_policy<reference_existing_object>());
    def("extract_X_ref", extract_X_ref, return_value_policy<reference_existing_object>());

    def("check_bool", check_bool);
    def("check_list", check_list);
    def("check_cstring", check_cstring);
    def("check_string", check_string);
    def("check_string_cref", check_string_cref);
    def("check_X", check_X);
    def("check_X_ptr", check_X_ptr);
    def("check_X_ref", check_X_ref);

    def("double_X", double_X);

    def("count_Xs", &X::count);
        ;

    object x_class(
        class_<X>("X", init<int>())
            .def( "__repr__", x_rep));
        
    // Instantiate an X object through the Python interface
    object x_obj = x_class(3);

    // Get the C++ object out of the Python object
    X const& x = extract<X&>(x_obj);
    if (x.value() != 3) {
        throw std::runtime_error("x.value() == 3 failure.");
    }
}


#include "module_tail.cpp"

File: external/boost_1.55.0/libs/python/test/extract.py
# Copyright David Abrahams 2004. Distributed under the Boost
# Software License, Version 1.0. (See accompanying
# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
'''
    >>> from extract_ext import *

Just about anything has a truth value in Python

    >>> assert check_bool(None)
    >>> extract_bool(None)
    0

    >>> assert check_bool(2)
    >>> extract_bool(2)
    1

    >>> assert not check_bool('')

Check that object manager types work properly. These are a different
case because they wrap Python objects instead of being wrapped by them.

    >>> assert not check_list(2)
    >>> try: x = extract_list(2)
    ... except TypeError, x:
    ...     if str(x) != 'Expecting an object of type list; got an object of type int instead':
    ...         print x
    ... else:
    ...     print 'expected an exception, got', x, 'instead'

Can't extract a list from a tuple. Use list(x) to convert a sequence
to a list:
    
    >>> assert not check_list((1, 2, 3))
    >>> assert check_list([1, 2, 3])
    >>> extract_list([1, 2, 3])
    [1, 2, 3]

Can get a char const* from a Python string:

    >>> assert check_cstring('hello')
    >>> extract_cstring('hello')
    'hello'

Can't get a char const* from a Python int:
    
    >>> assert not check_cstring(1)
    >>> try: x = extract_cstring(1)
    ... except TypeError: pass
    ... else:
    ...     print 'expected an exception, got', x, 'instead'

Extract an std::string (class) rvalue from a native Python type

    >>> assert check_string('hello')
    >>> extract_string('hello')
    'hello'

Constant references are not treated as rvalues for the purposes of
extract:

    >>> assert not check_string_cref('hello')

We can extract lvalues where appropriate:

    >>> x = X(42)
    >>> check_X(x)
    1
    >>> extract_X(x)
    X(42)

    >>> check_X_ptr(x)
    1
    >>> extract_X_ptr(x)
    X(42)
    >>> extract_X_ref(x)
    X(42)

Demonstrate that double-extraction of an rvalue works, and all created
copies of the object are destroyed:

    >>> n = count_Xs()
    >>> double_X(333)
    666
    >>> count_Xs() - n
    0

General check for cleanliness:

    >>> del x
    >>> count_Xs()
    0
'''

def run(args = None):
    import sys
    import doctest

    if args is not None:
        sys.argv = args
    return doctest.testmod(sys.modules.get(__name__))
    
if __name__ == '__main__':
    print "running..."
    import sys
    status = run()[0]
    if (status == 0): print "Done."
    sys.exit(status)
